keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1387 @@
+Kubelet (or another Container Orchestrator) to the container runtime to expand
+the file system.
+```
+message NodeExpandVolumeResponse {
+  ...
+  // Source device (e.g. /dev/sdX) whose mount is deferred to a container runtime
+  // that is capable of expanding the filesystems mounted by it.
+  // This field is OPTIONAL.
+  // It SHOULD be populated by a SP if runtime_supports_expand was set in
+  // NodeExpandVolumeRequest and SP is capable of deferring filesystem mount
+  // and expand requests to the container runtime.
+  // SP MUST populate `capacity_bytes` field with the desired capacity if source
+  // is specified indicating deferral of expansion to the container runtime.
+  string source = 2; [(alpha_field) = true];
+}
+```
+
+#### Enhancements in CRI Mount message
+
+The CRI `Mount` message (passed as part of `ContainerConfig` in
+`CreateContainerRequest`) needs to be enhanced to pass the file system type and
+mount options to be used by a container runtime as part of executing the deferred
+mount from a CSI plugin.
+
+The name of the existing `host_path` field is changed to a more generic `source`.
+This aligns better with the goal of being to support runtime assisted mounting for
+block devices from the host as well as remote file systems which may not have a
+path in the host as they can be directly mounted by the container runtime inside
+the sandbox. The same reasoning applies for not adding the `type` and `mount`
+fields to the `Device` message which is used to pass raw block device information
+to the container runtime.
+
+```
+// Mount specifies a host volume to mount into a container.
+message Mount {
+    ...
+    // Source can be a path in the host (for regular bind-mount
+    // scenarios), a block device or a NFS or CIFS server with
+    // exported path (for deferred mount scenarios). If the
+    // hostPath doesn't exist, then runtimes should report
+    // error. If the hostpath is a symbolic link, runtimes
+    // should follow the symlink and mount the real destination
+    // to the container.
+    string source = 2;
+    ...
+    // Type of the filesystem to mount: either bind or xfs, ext4,
+    // nfs, ntfs, etc. for deferred mount scenarios
+    string type = 6;
+    // Mount options corresponding to the filesystem to be mounted
+    map<string, string> options = 7;
+}
+```
+
+The new CRI fields: `type` and `options` will be populated by Kubelet (based on
+`NodePublishVolumeResponse` from the CSI plugin). Entries in these fields will be
+passed by a CRI runtime to the OCI runtime using the OCI [mounts](https://github.com/opencontainers/runtime-spec/blob/master/config.md#mounts) field.
+
+Note that the OCI spec already allows the `type` and `options` fields to be
+specified as part of the `mount` field. Therefore, no enhancements should be
+necessary in the OCI spec. OCI runtimes supporting mount deferral need to be
+able to execute the filesystem mount with the details specified in `mounts`.
```

Comment:
How can this be checked? Do users have to ensure that they don't ask for something that isn't supported, like creating a volume with XFS as filesystem and then using that with a runtime that doesn't support XFS?